signac/contrib/utility.py

```diff
@@ -479,3 +481,255 @@ def _nested_dicts_to_dotted_keys(t, encode=_encode_tree, key=None):
+
+    """
+    if path is None:
+        # Generate a path function based on the schema detected for jobs.
+        path_function = _make_schema_based_path_function(jobs=jobs)
+
+    elif path is False:
+
+        # Just use the job id as path.
+        def path_function(job):
+            """Use job id to construct path.
+
+            Parameters
+            ----------
+            job : :class:`~signac.contrib.job.Job`
+                An instance of :class:`~signac.contrib.job.Job`.
+
+            Returns
+            -------
+            str
+                Job id.
+
+            """
+            return str(job.id)
+
+    elif isinstance(path, str):
+        # Detect keys that are already provided as part of the path specifier
+        # and should therefore be excluded from the 'auto'-part.
+        exclude_keys = [x[1] for x in Formatter().parse(path)]
+
+        # Generate the paths based on the provided selection, excluding any keys
+        # that have already been provided by the user.
+        paths = _make_schema_based_path_function(jobs=jobs, exclude_keys=exclude_keys)
+
+        def path_function(job):
+            """Format a path based on a given string and schema.
+
+            Parameters
+            ----------
+            job : :class:`~signac.contrib.job.Job`
+                An instance of :class:`~signac.contrib.job.Job`.
+
+            Returns
+            -------
+            str
+                Formatted path.
+
+            """
+            try:
+                try:
+                    ret = path.format(job=job, **job.statepoint)
+                except TypeError as error:
+                    if (
+                        str(error)
+                        == "format() got multiple values for keyword argument 'job'"
+                    ):
+                        try:
+                            ret = path.format(job=job)
+                        except KeyError:
+                            raise _SchemaPathEvaluationError(
+                                "You must use fully qualified fields for this path, because the "
+                                "state point contains a key called 'job', e.g.: '{job.sp.job}' "
+                                "instead of '{job}'."
+                            )
+                    ret = path.format(job=job)
+                return _AutoPathFormatter(paths).format(ret, auto=job)
+            except AttributeError as error:
+                raise _SchemaPathEvaluationError(error)
+            except KeyError as error:
+                raise _SchemaPathEvaluationError(f"Unknown key: {error}")
+            except Exception as error:
+                raise _SchemaPathEvaluationError(error)
+
+        # Check that the user-specified path generates a 1-1 mapping
+        _check_path_function(jobs, path_spec=path, path_function=path_function)
+
+    else:
+        raise ValueError(
+            "The path argument must either be `None`, `False`, or of type `str`."
+        )
```

Comment:
This disagrees with the docstring that says it accepts a `callable`. Based on my suggestion above, we should probably let this function pass through a callable and validate all possible path function values with `_check_path_function` at the end before returning.